File: built-in plugins dsp/mb-organ-reed.cpp
/**
 * MB Reed Organ
 * Category : instrument
 * Type     : organ
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage reed organ (harmonium) with breathy tone
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ORGAN_REED_H
#define MB_ORGAN_REED_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbOrganReed : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-organ-reed";
    static constexpr const char* PLUGIN_NAME    = "MB Reed Organ";
    static constexpr const char* PLUGIN_TYPE    = "organ";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bellows = 0.6f;  // range [0, 1]
    float stops = 0.5f;  // range [0, 1]
    float tremulant = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbOrganReed() = default;
    ~MbOrganReed() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bellows = std::clamp(params.bellows, 0f, 1f);
        params.stops = std::clamp(params.stops, 0f, 1f);
        params.tremulant = std::clamp(params.tremulant, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Reed Organ
        return input;
    }
};

#endif // MB_ORGAN_REED_H

File: built-in plugins dsp/mb-organ-reed.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbOrganReedPlugin: PluginDefinition = {
  id: "mb-organ-reed",
  slug: "mb-organ-reed",
  name: "MB Reed Organ",
  category: "instrument",
  type: "organ" as any,
  version: "1.0.0",
  description: "Vintage reed organ (harmonium) with breathy tone",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.4 },
    { type: "square", detune: 5, gain: 0.3 },
    { type: "noise", detune: 0, gain: 0.1 },
    { type: "sine", detune: 0, gain: 0.2 },
  ],
  envelope: { attack: 0.08, decay: 0.15, sustain: 0.9, release: 0.15 },
  parameters: [
    {
      id: "bellows",
      name: "Bellows",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "stops",
      name: "Stops",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tremulant",
      name: "Tremulant",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { bellows: 0.6, stops: 0.5, tremulant: 0.3, volume: 0.8 },
};

export default MbOrganReedPlugin;
